### DISCUSSION TITLE
Having trouble with extra render on clientside routing

### DISCUSSION THREAD

**Discussion by pkellner** [vercel/next.js#19173] [Help]
I have the following code that every time onClick gets executed, the page gets rendered twice. I want the URL to update silently which is why I have the shallow: true set, but still, it renders twice.  When I remove the push, the page renders just once correctly.
```
      <a
        onClick={() => {
          paginationDataVar({
            ...paginationData,
            currentPage: lastPage,
          });
          router.push(`/?page=${lastPage+1}`, undefined, {
            shallow: true,
          });
        }}
        className="page-link"
        aria-label="Last"
      >
        <i className="fa fa-angle-double-right" aria-hidden="true"></i>
      </a>
```

**Comment by jamesmosier** [comment]
`shallow: true` will always cause a rerender because internally Next is updating the routers Context/state to keep track of the apps current route. Therefore shallow is not just an update the browsers url but also an update to `next/router`‘s state.
